Implement Contract Verification

pynchmeister opened the pull request:
## Description 

Contract verification now exists at within the ```apps/verify``` project directory. Allows users to provide a contract address, source code, and metadata files in order to verify smart contracts using the Sourcify API.

## Visual

<img width="817" alt="Screenshot 2023-06-12 at 5 45 44 PM" src="https://github.com/modularcloud/explorer/assets/33232379/271ab899-3ceb-4f88-8cdb-c4db6d6030d2">

https://github.com/modularcloud/explorer/assets/33232379/50884e76-93d8-46c2-8a9b-f88e73d20745

## Future Work

May include displaying the contract verification on the same page as applicable contract addresses.

lzrscg commented on apps/web/pages/api/contract-verification/prisma/contract-verification.ts:
```diff
@@ -0,0 +1,58 @@
+  req: NextApiRequest,
+  res: NextApiResponse
+) {
+  const { contractAddress, chainId, files, uploadedFilesFolderUrl } = req.body;
```
how do we determine the chainId?

lzrscg commented on apps/web/pages/api/contract-verification/prisma/lib/prisma.ts:
```diff
@@ -0,0 +1,18 @@
+import { PrismaClient } from '@prisma/client';
```
I see a lot of files are inside the `/api` directory. I believe this should only be reserved for routes. Can you move the non-API routes elsewhere?

lzrscg commented on apps/web/tsconfig.json:
```diff
@@ -1,10 +1,26 @@
 {
   "extends": "tsconfig/nextjs.json",
```
can you try extending this base definition? I believe there is a package for this in the monorepo:
`/packages/tsconfig`

pynchmeister replied:
sure thing

lzrscg commented on babel.config.js:
```diff
@@ -0,0 +1,6 @@
+module.exports = {
```
can you explain why this was necessary? I don't have an issue with it, just want to understand

pynchmeister replied:
seems unnecessary, so it has been removed

lzrscg commented on package.json:
```diff
@@ -16,12 +16,26 @@
     "format": "prettier --write \"**/*.{ts,tsx,md}\""
   },
   "devDependencies": {
+    "@babel/preset-typescript": "^7.21.5",
```
can you install all dependencies in the `web` package.json instead

`npm i mypackage --workspace=web`

lzrscg commented on packages/tailwind-config/tailwind.config.js:
```diff
@@ -1,6 +1,7 @@
 const { fontFamily } = require("tailwindcss/defaultTheme");
 
 module.exports = {
+  mode: "jit",
```
also curious why this was necessary?

Aswanth-c replied:
It's better to enable tailwind jit 

The build time was slow and enabling jit helped in reducing the build time

You can read more here 

https://v2.tailwindcss.com/docs/just-in-time-mode

lzrscg replied:
I thought this was enabled by default?

mhstom commented on apps/web/pages/api/file-upload/generateurl.ts:
```diff
@@ -0,0 +1,17 @@
+import { NextApiRequest, NextApiResponse } from "next";
+import generateUploadUrl from "./s3file-upload";
+
+export default async function generateS3Url(
```
who is using this?

mhstom commented on apps/web/ui/contract-verify/verifyAndUpload.tsx:
```diff
@@ -0,0 +1,202 @@
+        data.uploadedFilesFolderUrl = extractedFolderUrl
+          .slice(0, extractedFolderUrl.length - 1)
+          .join("/");
+        const uploadImage = await axios.put(getImageUploadUrl.data.url, file, {
```
could we compress the file before uploading?

mhstom reviewed (commented):
For the upload to s3 logic, could we compress the file before uploading?

lzrscg commented on apps/web/jest.config.js:
is this change still necessary?

lzrscg commented on apps/web/package.json:
```diff
@@ -6,6 +6,7 @@
     "dev": "next dev",
     "build": "next build",
     "start": "next start",
+    "postinstall": "prisma generate",
```
is this change still necessary?

pynchmeister replied:
No, this is unnecessary as it was moved to the ```verify``` project

lzrscg commented on apps/web/pages/_app.tsx:
```diff
@@ -0,0 +1,8 @@
+import '../styles/globals.css';
```
is this change still necessary?

lzrscg commented on apps/web/tsconfig.json:
```diff
@@ -1,10 +1,26 @@
 {
   "extends": "tsconfig/nextjs.json",
+  "compilerOptions": {
```
is this change still necessary?

mhstom commented on apps/verify/contract-verify/pages/verifyAndUpload.tsx:
```diff
@@ -0,0 +1,202 @@
+            }
+            zip.generateAsync({ type: "blob" }).then(async (content) => {
+              const zipFile = new Blob([content], { type: "application/zip" });
+              await uploadFile(zipFile);
```
if I am reading the code correctly, we are uploading files to s3 BEFORE sending the file to sourcify. Is that intentional? Any reason to store the files before they are verified?

mhstom replied:
if I am reading the code correctly, we are uploading files to s3 BEFORE sending the file to sourcify for verification. Is that intentional? Any reason to store the files before they are verified?

lzrscg commented on packages/@modularcloud/design-system/components/contract-verify/index.tsx:
```diff
@@ -0,0 +1,12 @@
+export function ContractVerify() {
```
no need to change this, but we are deprecating `@modularcloud/design-system`

this can just go the verify project folder

lzrscg commented on apps/verify/contract-verify/prisma/lib/prisma.ts:
```diff
@@ -0,0 +1,18 @@
+let prisma: PrismaClient;
+
+if (process.env.NODE_ENV === 'production') {
+  prisma = new PrismaClient();
```
does this need an environment variable?

lzrscg reviewed (changes requested):
This looks really good!

- Please ensure minimal changes occur outside of the `verify` directory.
- Why did you call it `verify/contract-verify` instead of just `verify` or `contract-verify`? It's not a problem, but just curious?

mhstom commented on apps/verify/contract-verify/pages/api/contract-verification/contract-verification.ts:
```diff
@@ -0,0 +1,67 @@
+        const record = await prisma.verification.create({
+          data: {
+            verificationStatus:
+              response.data.result[0].status == "perfect" ? "FULL" : "PARTIAL",
```
what do FULL and PARTIAL mean?

mhstom replied:
what do FULL and PARTIAL mean? When would the result be PARTIAL?

pynchmeister commented:
> * Why did you call it `verify/contract-verify` instead of just `verify` or `contract-verify`? It's not a problem, but just curious?

I moved the folder manually and it was called ```verify/contract-verify``` originally. But it is much cleaner as ```verify``` so I've switched it to that.

Aswanth-c commented on apps/web/package.json:
we can get rid of the changes here
as we wont be using it
